### ISSUE TITLE
Rule proposal: Disallow enums

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#8430]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I think `enum` should never be used. There are many reasons and a lot of articles explaining why, and I'm surprised that there is no rule against them and I can't find an open issue (I searched `enum in:title` and it showed a hundred closed issues 🥲).

https://dev.to/ivanzm123/dont-use-enums-in-typescript-they-are-very-dangerous-57bh

even the official docs list numerous pitfalls and suggest using a plain `object as const` instead:

https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls

> The biggest argument in favour of this format over TypeScript’s enum is that it keeps your codebase aligned with the state of JavaScript, and when/if enums are added to JavaScript then you can move to the additional syntax.

### Fail Cases

```TypeScript
enum Roles {
  Admin,
  Writer,
  Reader
}
```

### Pass Cases

```TypeScript
type Roles = "admin" | "writer" | "reader";
```

or the similar but unnecessary:

```ts
const Roles = {
  Admin: "admin",
  Writer: "writer",
  Reader: "reader"
} as const;
```

### Additional Info

Editor suggestions still work with plain types, without having to import anything:

<img width="679" alt="Screenshot" src="https://github.com/typescript-eslint/typescript-eslint/assets/1402241/e9fd3835-ad54-4407-acc9-c17c4f4d3df7">

**Comment by bradzacher** [comment]
https://typescript-eslint.io/linting/troubleshooting#how-can-i-ban-specific-language-feature

Even has an example of banning enums!

---

I disagree with the articles and believe they spread FUD.
A lot of the reasons against enums are founded in the use of number enums which are bad and unsafe. Just don't use number enums and they're a useful enough construct.

Manually defining a pseudo-enum just isn't a good option and it leads to your previously filed problems of anonymous types (except they're anonymous strings not objects) - so refactorings don't always work on them for example.

But if it's your choice you're free to go that route with the above FAQ.

**Comment by fregante** [comment]
> they spread FUD

Including TypeScript's own documentation?

**Comment by bradzacher** [comment]
> Including TypeScript's own documentation? 

The section you linked is specifically titled: "Const enum pitfalls". Not "Enum pitfalls" - just about const enums - a subset of the space.

But yes - there's some FUD there - for example point 2 only applies to libraries, not application code and point 3 is mostly moot in codebases I've seen (yet to see an app codebase using that compiler option).

> but that covers a lot of existing rules as well.

In the past we were happy to accept rules that just ban syntax and did nothing more - we were young and looking to grow as a project. Then we got big enough and realised we didn't want to maintain a thousand tiny rules.

We also realised that with such rules existing it also in a way acted as us endorsing such patterns and as such caused people to want options in rules to do the opposite and then people would be annoyed by defaults etc. So a lot of rules we either morphed into larger rules (like `consistent-object-style` andn`naming-convention`) or we removed them.

Now our stance really is "if a proposal is just a selector that reports - that's a job for `no-restricted-syntax`, not a bespoke rule".
